### ISSUE TITLE
Advertise deprecated endpoints via Deprecation headers ( draft-ietf-httpapi-deprecation-header-03 )

### ISSUE THREAD

**Issue by ThisIsMissEm** [mastodon/mastodon#30308]
### Pitch

Currently notices of deprecated endpoints are exclusively listed in that endpoints' documentation Version History. This means that people may not notice that they are using deprecated endpoints, and API Clients can't automatically warn when a deprecated API is used.

I'd like to propose that we adopt [draft-ietf-httpapi-deprecation-header-03](https://datatracker.ietf.org/doc/html/draft-ietf-httpapi-deprecation-header-03), and advertise as follows:

```text
Deprecation: @1688169599
Link: <https://docs.joinmastodon.org/api/deprecations#<id>; rel="deprecation"; type="text/html"
```

Where `1688169599` is when the endpoint was or will be deprecated at, in seconds; and `<id>` is a unique identified for that deprecation on the deprecation pages.

We may also have use in the `Sunset` HTTP Header, however, this would require fairly exact release dates, which we don't currently have, or we use the `sunset` rel in a Link header, to indicate upcoming removal of an API. (but I'm not sure if we've ever completely removed an API, a breaking change.

There is a [ticket on the documentation](https://github.com/mastodon/documentation/issues/1443) repository which is complementary to this.

### Motivation

This will help developers be better able to discover when they're using outdated API endpoints, and help them upgrade, ensuring a healthier API community usage.

**Comment by LucasGGamerM** [comment]
I am in favor of this

**Comment by VyrCossont** [comment]
Makes sense for client library developers and bot operators, who would be in a position to see error messages if their client library was aware of these header. I assume this isn't meant to be surfaced to end users.

**Comment by ThisIsMissEm** [comment]
> I assume this isn't meant to be surfaced to end users.

@VyrCossont correct! Though, if you're a developer using a client library, then the client library may choose to expose this information to you via logging.
